rfd/0083-license-expiration-warnings.md

```diff
@@ -0,0 +1,144 @@
+  Logins:             someUser
+  Valid until:        2022-08-16 03:38:07 +0100 IST [valid for 12h0m0s]
+  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty
+
+  Your Teleport Enterprise Edition license has expired. Please reach out to your
+  Teleport Account Manager (or licenses@goteleport.com) to obtain a new license.
+```
+
+```
+$ build/tctl status -c /home/edward/teleport/teleport.yaml
+Cluster  localhost                                                               
+Version  10.1.2                                                                  
+host CA  never updated                                                           
+user CA  never updated                                                           
+db CA    never updated                                                           
+jwt CA   never updated                                                           
+CA pin   sha256:3d72102f020146d09ff400810f59f70b8163ebfe6ec1ecfb0b3b2a0c151592
+
+Your Teleport Enterprise Edition license will expire in 10 days. Please reach out
+to   your Teleport Account Manager (or licenses@goteleport.com) to obtain a new
+license.
+```
+
+
+CLI does not support snoozing the warnings as you can in the webUI. Which provides the functionality to disable the warnings temporarily.
+
+### Web UI UX
+
+For web UI - banner on top
+Show warning (yellow) to users N days prior to expiration, 1 day snooze, based on license_warning_days role option with same warning text as the CLI
+Show error (red) to all users after expiration, no dismiss or snoozing available.
+Snoozing allows users to disable the warning in the web UI from being displayed for N days.
+
+### Configuration UX
+
+The time interval in which to show a warning before expiry can be configured on a per role basis.
+Interval can be configured by modifying a role option ‘license_warning_days’.
+Default will be 30 days, admins can configure if needed.
+Default preset roles will use the default. For example the editor/auditor/access roles.
+When multiple roles for the same user have different values for this interval the shortest one will be used.
+This value cannot be configured to be below the minimum value of 10 days.
+
+Idea: Keep this setting as a role option
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: editor
+spec:
+  options:
+    license_warning_days: 90  # ← defaults to 30, min 10
+```
+
+Default values for default roles:
+Editor - 90 days
+Auditor - 30 days
+Access - 30 days
+
+If you have multiple roles, the shortest time period takes precedence.
+
+## Implementation details
+
+Existing web api endpoint ‘/ping’ already returns a license warning. Due to this endpoint being unauthenticated and the proposed functionality requires the ability to differentiate based on identity/role this cannot be used as is.
+Instead this RFD proposes a new authenticated endpoint ‘/license’ be exposed to provide the required functionality. This would consist of
```

Comment:
Can you document the full reference for the new endpoint? The full path, request, response in a more structured way. Something like (but clean it up):

```
GET /webapi/...../check-license

Request: empty
Response: {"messages": ["..."]}
```

Reply:
FYI, @fspmarshall's work introduces a new backend resource called a "cluster alert" that is to be used to notify the user when an upgrade is available. @hatched is working on a new endpoint to expose these alerts.

I think a license expired or about to expire also qualifies as a cluster alert.

Can we look to combine these two approaches so we don't have two separate ways to present alerts to the user?

cc @hatched @r0mant @xinding33

Reply:
I'm not sure we'll be able to reuse the same mechanism because the license warnings will be different based on the identity of the user, users can have different intervals of when the warning will be shown to them. We'll see what we can reuse, if anything.

Reply:
Maybe we can reuse the same data structures and API at least, so if there will be an API for retrieving cluster alerts, the handler can also look at whether there are license warnings for a particular user who's making this request and enrich the "cluster alert" with the additional license warning before returning it to the user. This I think would make sense.

Reply:
@hatched Is there a PR for exposing these alerts to users which we can piggyback on? @EdwardDowling Let's update the RFD with this.

Reply:
The UI is actually using `GET /enterprise/license/status` not `GET /ping` to get this data. This endpoint is also unauthenticated.

Reply:
If it's unauthenticated, then we won't be able to use it to implement the requirements in this RFD.

Reply:
> I'm not sure we'll be able to reuse the same mechanism because the license warnings will be different based on the identity of the user, users can have different intervals of when the warning will be shown to them.

FWIW the "cluster alerts" decide if a user is allowed to see a given alert based on labels.  If the logic around which users should see which license warnings is fairly straightforward, you might just be able to add a new label and update `ServerWithRoles.GetClusterAlerts` to do some special checks if it sees that label.